api/rosetta/data_integration_test.go

```diff
@@ -0,0 +1,269 @@
+// Copyright 2021 Alvalor S.A.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+// +build integration
+
+package rosetta_test
+
+import (
+	"encoding/hex"
+	"runtime"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/dgraph-io/badger/v2"
+	"github.com/klauspost/compress/zstd"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/onflow/flow-go/model/flow"
+
+	"github.com/optakt/flow-dps/api/rosetta"
+	"github.com/optakt/flow-dps/models/dps"
+	"github.com/optakt/flow-dps/rosetta/configuration"
+	"github.com/optakt/flow-dps/rosetta/identifier"
+	"github.com/optakt/flow-dps/rosetta/invoker"
+	"github.com/optakt/flow-dps/rosetta/retriever"
+	"github.com/optakt/flow-dps/rosetta/scripts"
+	"github.com/optakt/flow-dps/rosetta/validator"
+	"github.com/optakt/flow-dps/service/dictionaries"
+	"github.com/optakt/flow-dps/service/index"
+	"github.com/optakt/flow-dps/testing/snapshots"
+)
+
+const (
+	invalidBlockchain = "invalid-blockchain"
+	invalidNetwork    = "invalid-network"
+	invalidToken      = "invalid-token"
+
+	invalidBlockHash = "af528bb047d6cd1400a326bb127d689607a096f5ccd81d8903dfebbac26afb2z" // invalid hex value
+
+	validBlockHashLen = 64
+)
+
+func setupDB(t *testing.T) *badger.DB {
+	t.Helper()
+
+	opts := badger.DefaultOptions("").
+		WithInMemory(true).
+		WithReadOnly(true).
+		WithLogger(nil)
+
+	db, err := badger.Open(opts)
+	require.NoError(t, err)
+
+	reader := hex.NewDecoder(strings.NewReader(snapshots.Rosetta))
+	dict, _ := hex.DecodeString(dictionaries.Payload)
+
+	decompressor, err := zstd.NewReader(reader,
+		zstd.WithDecoderDicts(dict),
+	)
+	require.NoError(t, err)
+
+	err = db.Load(decompressor, runtime.GOMAXPROCS(0))
+	require.NoError(t, err)
+
+	return db
+}
+
+func setupAPI(t *testing.T, db *badger.DB) *rosetta.Data {
+	t.Helper()
+
+	index := index.NewReader(db)
+
+	params := dps.FlowParams[dps.FlowTestnet]
+	config := configuration.New(params.ChainID)
+	validate := validator.New(params, index)
+	generate := scripts.NewGenerator(params)
+	invoke, err := invoker.New(index)
+	require.NoError(t, err)
+	retrieve := retriever.New(params, index, validate, generate, invoke)
+	controller := rosetta.NewData(config, retrieve)
+
+	return controller
+}
```

Comment:
As Max suggested last week, we could start making those helpers return errors instead and run the `require.NoError` statements from the caller. I feel like the way it is here is nicer for clarity though, since it avoids having to check the error in each test. We can always just assume that it's all good.

WDYT is better?

Reply:
I usually like being clear and/or explicit, so I like the idea of helper functions returning errors.
However, since these are executed at the start of each test case, before any other code is ran, for me mentally they are part of the test case setup process, so I'm fine with them being as they are.

For instance, I have since refactored the code that prepares the http request/recorder/echo context for each test case into a `setupRecorder` helper function (part of another PR). For that I used the function call that returned an error and `require`-d that it did not fail, since it's part of the test case execution and not part of the initial scaffolding.